### ISSUE TITLE
Extending FOSUserBundle alternatives

### ISSUE THREAD

**Issue by pribeirojtm** [FriendsOfSymfony/FOSUserBundle#2734]
**[DISCUSSION]**
Hi,

I'm using sf 3.4 and currently I have a UserBundle that extends FosUserBundle.
```php
//src/UserBundle/UserBundle.php
namespace UserBundle;

use Symfony\Component\HttpKernel\Bundle\Bundle;

class UserBundle extends Bundle
{
    /**
     * @return string
     */
    public function getParent()
    {
        return 'FOSUserBundle';
    }

}
```
I'm doing this because I extends and adapt many twig templates, adapting some layout and so.
Also, I extend some controllers to add extra /change base functionality. 

Example:

```php
<?php

//src/UserBundle/Controller/RegistrationController.php
namespace UserBundle\Controller;

use FOS\UserBundle\Form\Factory\FactoryInterface;
use Symfony\Component\EventDispatcher\EventDispatcherInterface;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\Security\Core\Exception\AccessDeniedException;
use FOS\UserBundle\Model\UserInterface;
use FOS\UserBundle\Model\UserManagerInterface;
use FOS\UserBundle\Controller\RegistrationController as BaseRegistrationController;
use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface;

use Sensio\Bundle\FrameworkExtraBundle\Configuration\Route;


class RegistrationController extends BaseRegistrationController
{

    protected $userManager;

//NOTE: WITH recent changes I must inject parent needed Services and call parent construct ....
//Code below is the code working without recent changes in FosUserBundle
    public function __construct(UserManagerInterface $userManager)
    {
        $this->userManager = $userManager;
    }


    /**
     * Tell the user his account is now confirmed.
     *
     * ADDING the functionality to add default group to the user during registration by overriding the confirmAction in Registration Controller
     *
     * It works like "fos_user_registration_confirmed", but it is costumized
     *
     * @Route("/register/confirmed", name="user_registration_confirmed")
     * @return Response
     * @throws \LogicException
     * @throws \Symfony\Component\Security\Core\Exception\AccessDeniedException
     */
    public function confirmedAction(Request $request)
    {
        //$em = $this->getDoctrine()->getManager();

        /** @var \AppBundle\Entity\User $user */
        $user = $this->getUser();

        if (!\is_object($user) || !$user instanceof UserInterface) {
            throw new AccessDeniedException('This user does not have access to this section.');
        }

        //$userManager = $this->get('fos_user.user_manager');//avoid. Inject service instead
        $userManager = $this->userManager;

        /**
         * @var \AppBundle\Repository\GroupsRepository $repository
         */
        $repository = $this->getDoctrine()->getRepository('AppBundle:Groups');
        /** @noinspection PhpUndefinedMethodInspection */
        $group = $repository->findOneByName('staff');//group named 'staff' must exists!
        $user->addGroup($group);
        //$user->addRole('ROLE_ADMIN');

        $userManager->updateUser($user);

        /* // NOTE: code in the parent action...
        if (!is_object($user) || !$user instanceof UserInterface) {
            throw new AccessDeniedException('This user does not have access to this section.');
        }

        return $this->render('FOSUserBundle:Registration:confirmed.html.twig', array(
            'user' => $user,
        ));*/

        return parent::confirmedAction($request);

    }

}
```

Do you recommend another way of using FosUserBundle?

Thanks in advance

**Comment by stof** [comment]
Well, there are different things here:

- Symfony supports overriding bundle templates at the project-level instead of using a child bundle (which is deprecated in Symfony 3.4 anyway). So put your template in `app/Resources/FOSUserBundle/views` to override the bundle ones (the path is different when using Flex, just read the Symfony documentation about it).
- to add groups automatically on registration, you should rather hook into events of the bundle. Your implementation even implements this on the wrong action (it does not do it in the POST request doing the registration, but on the GET page displayed after the redirection, breaking the HTTP semantic by making it perform changes). For your case, you should put your `$user->addRole()` logic in a listener on the `\FOS\UserBundle\FOSUserEvents::REGISTRATION_SUCCESS` event (don't call `$userManager->updateUser($user)` in your listener, as the controller does it already)

**Comment by pribeirojtm** [comment]
Thanks for the tips @stof .
I'm currently refactoring some parts of my UserBundle, and I want ultimately to remove this inheritance from FOSUserBundle.

Forgot to say before that I also have other "extends" (dependencies) applied besides the ones mentioned above (twig templates and Controllers) .I have PromoteUserCommand and DemoteUserCommand extending the bases Controllers changing description of the commands, and have Util\UserManipulator also extending the base one to change /add functionality to addRole(),promote(), demote(), removeRole() methods. I think there aren't other inheritance dependencies in the bundle. 
So I guess this will be my main problem to solve. Can you advice me on this?

I've already solved the problem with twig templates using overriding bundle templates at the project-level instead of using a child bundle. Regarding Controllers I'm trying to apply your suggestions using event listeners. I have also a form in ChangePasswordController inside changePasswordAction that is beeing changed to add a new extra hidden field to it, but I guess I can overtake the problem following the recommendations in this link: [https://symfony.com/doc/master/bundles/FOSUserBundle/overriding_forms.html](url)

Thanks for your help.

**Comment by stof** [comment]
> I have PromoteUserCommand and DemoteUserCommand extending the bases Controllers changing description of the commands, and have Util\UserManipulator also extending the base one to change /add functionality to addRole(),promote(), demote(), removeRole() methods. I think there aren't other inheritance dependencies in the bundle.
> So I guess this will be my main problem to solve. Can you advice me on this?

Well, if you need custom logic to promote or demote roles, the simpler solution is probably to override the commands entirely (by defining your own commands with the same names)., defining your own command from scratch with your own logic (this could even remove the need to extend UserManipulator, as it is used only by the commands)

**Comment by pribeirojtm** [comment]
Thanks again stof.

I've refactored my application to remove this inheritance bundle dependency, that it is not recommended to do it since sf3.4. I understand the benefits from that way of programming, but this way it has some limitations I guess. 

When I was refactoring, I have the need of apply some changes when, for example, the user fails changing its password. I did not find any event to use like _CHANGE_PASSWORD_FAILED_ to use in an event listener. How do I handle those kind of situations? I wan't to change redirection, change flash message, or do something else in this case. I'm unable to customize this situation :\ Or do I have other way to accomplish that?

Regarding the commands, I have did not override **fos:user:demote** and **fos:user:promote** because I'm not extending their base commands. In my case I've wrote my own commands, like: **app:user:demote** and **app:user:promote** , _DemoteUserCommand_ and _PromoteUserCommand_, respectively and I had to create a new _AbstractRoleCommand_,that is extended by those 2 commands, with the code being exactly the base command with the exception of one line: 
_use FOS\UserBundle\Util\UserManipulator;_ turned into 
_use UserBundle\Util\UserManipulator;_ and I had to create my own _UserManipulator_ costumized.

Is this the (best) way for "extending" (without extending) some bundle functionalities?

**Comment by pribeirojtm** [comment]
Sorry, I forgot to close this issue. 
Extending this bundle is unnecessary and adds more limitations in programming by coupling the code and not allowing to upgrade to a higher sf version (sf4). 

I had the need to create custom commands like: **app:user:demote** and **app:user:promote** , _DemoteUserCommand_ and _PromoteUserCommand_, respectively and I had to create a new _AbstractRoleCommand_,that is extended by those 2 commands, with the code being exactly the base command with the exception of one line: 
_use FOS\UserBundle\Util\UserManipulator;_ turned into 
_use UserBundle\Util\UserManipulator;_ and I had to create my own _UserManipulator_ costumized.

@MWMathieu , regarding "overriding" bundle templates, I've solved it creating them (with the same name, following the same directory structure) at the project-level instead of using a child bundle, just for the sake of coeherences, since I'm using some routes from FosUserBundle, and some from my UserBundle (they are in app/Resources/FOSUserBundle/views/{ChangePawword/Profile/Group/Registration/Resetting/Security}/***.html.twig
